### PR TITLE
Initial commit for serverless pattern

### PR DESCRIPTION
**Pull request by diwakarkvs**
*Issue #, if available:* New serverless pattern - apigw-authorizer-sqs

*Description of changes:* This queue levelling pattern can be used when customers want to use APi Gateway RestAPI to send messages to Amazon SQS and have APIGW authorize the request.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: cdk-apigw-authorizer-sqs/README.md-OLD [hunk 1]
```diff
@@ -0,0 +1,17 @@
+## My Project
```
Comment (mavi888) [inline cdk-apigw-authorizer-sqs/README.md-OLD:1]: please remove this file

File: cdk-apigw-authorizer-sqs/README.md [hunk 2]
```diff
@@ -0,0 +1,60 @@
+# AWS Service 1 to AWS Service 2
```
Comment (mavi888) [inline cdk-apigw-authorizer-sqs/README.md:1]: fix title for readme

File: cdk-apigw-authorizer-sqs/README.md [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+This pattern << explain usage >>
+
+Learn more about this pattern at Serverless Land Patterns: << Add the live URL here >>
```
Comment (mavi888) [inline cdk-apigw-authorizer-sqs/README.md:5]: whole readme is missing

**Comment by mavi888** [comment]
@diwakarkvs please fill the README.md file correctly and make sure that there are no extra files commited in ths PR. 

After that is fixed i will continue my review. 

Cheers

**Comment by diwakarkvs** [comment]
@mavi888 , I updated the README.md and deleted README copy.md and READEME.md-OLD files. 
I had to add a few pictures to provide context to README content. So i checked in 3 images to /images folder.

File: cdk-apigw-authorizer-sqs/README.md [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+Important: this application uses various AWS services and there are costs associated with these services after the Free Tier usage - please see the [AWS Pricing page](https://aws.amazon.com/pricing/) for details. You are responsible for any AWS costs incurred. No warranty is implied in this example.
+
+## Requirements
+
+* [Create an AWS account](https://portal.aws.amazon.com/gp/aws/developer/registration/index.html) if you do not already have one and log in. The IAM user that you use must have sufficient permissions to make necessary AWS service calls and manage AWS resources.
+* [AWS CLI](https://docs.aws.amazon.com/cli/latest/userguide/install-cliv2.html) installed and configured
+* [Git Installed](https://git-scm.com/book/en/v2/Getting-Started-Installing-Git)
+* [CDK Installed](https://docs.aws.amazon.com/cdk/v2/guide/getting_started.html) 
+* [Create Lambda authorizer from blueprint](https://github.com/awslabs/aws-apigateway-lambda-authorizer-blueprints/blob/master/blueprints/python/api-gateway-authorizer-python.py)
```
Comment (mavi888) [inline cdk-apigw-authorizer-sqs/README.md:28]: why this cannot be created as IaC?

File: cdk-apigw-authorizer-sqs/README.md [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+## Deployment Instructions
+
+1. Create a new directory, navigate to that directory in a terminal and clone the GitHub repository:
+```
+git clone https://github.com/aws-samples/pattern-apigw-auth-sqs.git
+```
+2. Change directory
+```
+cd pattern-apigw-auth-sqs
```
Comment (mavi888) [inline cdk-apigw-authorizer-sqs/README.md:41]: this is not the right folder for this pattern

File: cdk-apigw-authorizer-sqs/README.md [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+* [Create Lambda authorizer from blueprint](https://github.com/awslabs/aws-apigateway-lambda-authorizer-blueprints/blob/master/blueprints/python/api-gateway-authorizer-python.py)
+
+![Alt text](images/lambda-auth-blueprint.png?raw=true "Create lambda function using blueprint")
+
+
+## Deployment Instructions
+
+1. Create a new directory, navigate to that directory in a terminal and clone the GitHub repository:
+```
+git clone https://github.com/aws-samples/pattern-apigw-auth-sqs.git
```
Comment (mavi888) [inline cdk-apigw-authorizer-sqs/README.md:37]: put the folder serverless-patterns

File: cdk-apigw-authorizer-sqs/README.md [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+* With Token Autorizer
+    ```
+    cdk synth ApigwSqsAuthStack -o ./cloudformation_templates
+    ```
+* Without a Token Autorizer
+    ```
+    cdk synth ApigwSqsStack -o ./cloudformation_templates
+    ```
+4. To deploy AWS resources as a CDK project
+   ##### Paramters needed for CDK
```
Comment (mavi888) [inline cdk-apigw-authorizer-sqs/README.md:53]: parameters

File: cdk-apigw-authorizer-sqs/README.md [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+    ```
+    cdk synth ApigwSqsAuthStack -o ./cloudformation_templates
+    ```
+* Without a Token Autorizer
+    ```
+    cdk synth ApigwSqsStack -o ./cloudformation_templates
+    ```
+4. To deploy AWS resources as a CDK project
+   ##### Paramters needed for CDK 
+    * storagequeue=Name of the SQS queue where you want data to be sent
```
Comment (mavi888) [inline cdk-apigw-authorizer-sqs/README.md:54]: what queue? is not in the prerequistes. Can you create it as Iac?

File: cdk-apigw-authorizer-sqs/README.md [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+    * awsregion=AWS region where you want the APIGW, SQS and token authoizer to be created.
```
Comment (mavi888) [inline cdk-apigw-authorizer-sqs/README.md:55]: can you not get this one from the cdk env?

File: cdk-apigw-authorizer-sqs/apigw_sqs/apigw_sqs_auth_stack.py [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+            description="Enter the AWS region name where the resources need to be deployed."
+        )
+        need_an_authorizer = core.CfnParameter(self, "authorizerneeded", type="String",
+                allowed_values= ["Yes"],
+                default="Yes",
+                description= "Do you need an Custom Authorizer? If Yes, enter a valid Lambda ARN value below.",
+    
+        )       
+        authorizer_lambda_arn = core.CfnParameter(self, "lambdaauthorizerarn", type="String",
+            default='arn:aws:lambda:us-east-1:764551143200:function:apigwRequestAuthorizer',
```
Comment (mavi888) [inline cdk-apigw-authorizer-sqs/apigw_sqs/apigw_sqs_auth_stack.py:30]: what account name is this one?

**Comment by mavi888** [comment]
This pattern is not yet ready for merging. 
Please fix the comments and then is the folder cloudformation templates needed? if not please remove.

Also please remove any extra files that is not needed for the project to work or is not a required one listed in here https://github.com/aws-samples/serverless-patterns/tree/main/_pattern-model. Such as contributing, code of conduct, or licence. 

is source.bat required? 

This project has an enourmous amount of files that im not sure they are required for this pattern to work. please clean

File: cdk-apigw-authorizer-sqs/README.md (hunk 5 above)
Comment (saikotag) [reply cdk-apigw-authorizer-sqs/README.md:41]: Done! updated the path to cdk-apigw-authorizer-sqs

File: cdk-apigw-authorizer-sqs/README.md (hunk 6 above)
Comment (saikotag) [reply cdk-apigw-authorizer-sqs/README.md:37]: Done! updated the path to https://github.com/aws-samples/serverless-patterns/cdk-apigw-authorizer-sqs.git

File: cdk-apigw-authorizer-sqs/README.md (hunk 7 above)
Comment (saikotag) [reply cdk-apigw-authorizer-sqs/README.md:53]: Done! corrected the spelling

File: cdk-apigw-authorizer-sqs/README.md (hunk 8 above)
Comment (saikotag) [reply cdk-apigw-authorizer-sqs/README.md:54]: The SQS queue will be created by IaC. This parameter will take a value for 'Queue Name' - Users can provide a user friendly name that will be created by IaC.

File: cdk-apigw-authorizer-sqs/apigw_sqs/apigw_sqs_auth_stack.py (hunk 10 above)
Comment (saikotag) [reply cdk-apigw-authorizer-sqs/apigw_sqs/apigw_sqs_auth_stack.py:30]: Deleted the default value

**Comment by saikotag** [comment]
Hi Mavi,
Updated the code and README.md file based on your comments and inputs. Please review it when you get a chance. Thanks

File: cdk-apigw-authorizer-sqs/README.md [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+
+1. Create a new directory, navigate to that directory in a terminal and clone the GitHub repository:
+```
+git clone https://github.com/aws-samples/serverless-patterns/cdk-apigw-authorizer-sqs.git
+```
+2. Change directory
+```
+cd cdk-apigw-authorizer-sqs
+```
+3. To generate a cloudformation templates (optional)
```
Comment (mavi888) [inline cdk-apigw-authorizer-sqs/README.md:42]: Can you take the region from the configuration and make CDK create the queue name. You can also use a config file inside the app to pass the parameters as this is not a clean way to do the call. 
Same for the deploying

File: cdk-apigw-authorizer-sqs/README.md [hunk 12]
```diff
@@ -0,0 +1,81 @@
+# CDK Python project for deploying Queue-Based Leveling pattern!
```
Comment (mavi888) [inline cdk-apigw-authorizer-sqs/README.md:1]: remove the exclamation mark

**Comment by mavi888** [comment]
Still some comments to fix in this PR.

Also why the authorizer function cannot be created as IaC in the project? The solution will be more elegant if that is added to it.

File: cdk-apigw-authorizer-sqs/README.md (hunk 12 above)
Comment (saikotag) [reply cdk-apigw-authorizer-sqs/README.md:1]: Done.

File: cdk-apigw-authorizer-sqs/README.md [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+```
+cd cdk-apigw-authorizer-sqs
+```
+3. To generate a cloudformation templates (optional)
+```
+aws_region_name=<aws_region_name> storage_sqs_queue_name=<user_friendly-SQS_Queue_name> authorizer_lambda_arn=<arn:aws:lambda:<aws_region>:<aws_account>:function:<lambda_function_name> cdk synth
+```
+4. To deploy AWS resources as a CDK project
+```
+aws_region_name=<aws_region_name> storage_sqs_queue_name=<user_friendly-SQS_Queue_name> authorizer_lambda_arn=<arn:aws:lambda:<aws_region>:<aws_account>:function:<lambda_function_name> cdk deploy ApigwSqsAuthStack
```
[the remark this answers is not in the document]
Comment (saikotag) [reply cdk-apigw-authorizer-sqs/README.md:48]: Done! Added instructions to import modules

File: cdk-apigw-authorizer-sqs/README.md [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+## How it works
+
+At the end of the deployment the CDK output will list stack outputs, and an API Gateway URL. In the customer's AWS account, a REST API along with an authorizer and a SQS queue will be created.
+
+![Alt text](images/apigw-aws-console.png?raw=true "AWS Console showing API Gateway creation")
+## Testing
+
+Test the api invocation using curl:
+```
+curl -X POST "https://<apigateway-id>.execute-api.us-east-1.amazonaws.com/prod/sqs" -H "Authorization:testtoken"  -H "Content-Type: application/json" --data '{"status": "ok"}'
```
[the remark this answers is not in the document]
Comment (saikotag) [reply cdk-apigw-authorizer-sqs/README.md:60]: Done.

**Comment by saikotag** [comment]
Since APIGW is using custom Lambda authorizer, where users can implement their own token validation logic, I feel that it is best to have users implement their own lambda logic that best suits them. Although, the lambda function can be created as  IaC they still have to customize and write their own code, instead as 'Requirements' we have a step
[Create Lambda authorizer from blueprint](https://github.com/awslabs/aws-apigateway-lambda-authorizer-blueprints/blob/master/blueprints/python/api-gateway-authorizer-python.py)

**Comment by mavi888** [comment]
This pattern was merged and published in this link: https://serverlessland.com/patterns/cdk-apigw-authorizer-sqs

Thanks a lot for contributing!

**Pull request merged**